### ISSUE TITLE
add ability to parse nested dictionary json?

### ISSUE THREAD

**Issue by hlung** [Baglan/MCLocalization#8]
I think it should also be able to read nested dictionary json. 

For example, this json...
```
{
  "greeting": "Hello!",
  "error": {
    "no-connection": "No connection",
    "bad-input": "Bad Input",
  }
}
```
We may use a dot syntax to accesss the nested dictionary like this...
`[MCLocalization stringForKey:@"error.no-connection"];`

**Comment by Baglan** [comment]
Hi,

Thank you for your suggestion!

You can already use arbitrary keys such as "error.no-connection", so, as far as I can see, the value of this approach would be to introduce hierarchical structure to the way strings are stored. Could you tell me why that additional complexity would be beneficial?

Regards,
–Baglan

**Comment by hlung** [comment]
Well, obviously having hierarchical structure would make the JSON file look more organized. The project I am working on has a quite big localization file, and it has hierarchical structure. So it would be nice to have.

**Comment by Baglan** [comment]
I see your point and I understand how that could be useful for your project.

I think using arbitrary delimiter in is not a good solution because you don't know how other people name their localization keys an imposing an arbitrary standard "post factum" could break existing projects. It would also make sense to support arbitrary depth of the hierarchy. Together, than means that for a key such as "domain.subdomain.key" you would need to, potentially, make  a lot of lookups in the data to determine which string to use of if that string even exists. That could also add some ambiguity in case both "domain / subdomain.key" and "domain / subdomain / key" exist.

One way to avoid multiple lookups would be to flatten data on load, although that doesn't solve the ambiguity. That is, also, more efficient to do once with something like a script when exporting localization to the project than doing it every time localization data is loaded.

Another possibility I've thought about would be to express the hierarchy in the API, say, with a "variadic" lookup function. Something like this:

	[MCLocalization.stringForKey:@"domain", @"subdomain", @"key"];

"Variadic" parameter, however, should be the last one in the parameter list and that would be a problem for the MCLocalization stringForKey:withPlaceholders: method.

So far, I think this feature, although, potentially, useful, would, on balance, be too much trouble to be worth implementing. I'll think about it for a few more days – perhaps I can find a better solution that would also be backwards compatible.

In summary:
- If I was in your situation, to solve this problem, I would write a script to flatten the hierarchy**; 
- If you find a solution that would, ideally, support arbitrary depth of hierarchy, avoid multiple lookups into the data, remove the possibility for ambiguity and have an API backwards compatible with the existing one, I would love to hear bout it;
- If you have a good argument for why hierarchy in localization data is fundamentally important, I would consider a spin-off project.

** In our project, we use YAML as a source for the localization data. YAML is much more readable, especially for multi-line strings. If you are going to process localization data with a script anyway, YAML, or some other source format, could be a good solution. You can easily convert existing JSON to YAML so that you don't have to re-type everything. We have an existing solution in ruby and I can share it with you if you consider something like that.

**Comment by Baglan** [comment]
Hi,

I decided not to add this feature, however, I've exposed some internal methods that should help you extend MCLocalization to support hierarchical data.

I also wrote a sample class that should do what you need and posted it as a gist here:

https://gist.github.com/Baglan/b898a659a71ca97a3d73

A more general solution would be to delegate searching for localization keys to MCLocalizationDataSource implementations, however, that would make the current API backward incompatible – I will consider it for a next major version or for a future spin-off.

Please let me know if you have further questions.

Regards,
–Baglan

**Comment by hlung** [comment]
Thanks! I'll give it a shot.

**Comment by hlung** [comment]
@Baglan btw, don't forget to push the .podspec file to the trunk!

**Comment by Baglan** [comment]
Done; thank you for reminding me!

**Comment by hlung** [comment]
Hello again! I ended up forking your project to support nested dictionary. Good news is, I only add one small change. :dancers: 

I replace
`NSObject * lookupResult = langugeStrings[key];`
with
`NSObject * lookupResult = [langugeStrings valueForKeyPath:key];`

Here's the definition of this handy `-valueForKeyPath:` method in NSKeyValueCoding.h 

/* Key-path-taking variants of like-named methods. The default implementation of each parses the key path enough to determine whether or not it has more than one component (key path components are separated by periods). If so, -valueForKey: is invoked with the first key path component as the argument, and the method being invoked is invoked recursively on the result, with the remainder of the key path passed as an argument. If not, the like-named non-key-path-taking method is invoked.
*/
- (id)valueForKeyPath:(NSString *)keyPath;

This however has a limitation that keys must be separated by periods "." But I think it's ok since it's quite normal NOT to include periods in keys.

Check it out [here](https://github.com/hlung/MCLocalization).
PS. I also updated your .gitignore. :)
